deltacat/aws/s3u.py

```diff
@@ -504,41 +511,87 @@ def download_manifest_entries_distributed(
 
 def upload(s3_url: str, body, **s3_client_kwargs) -> Dict[str, Any]:
 
-    # TODO (pdames): add tenacity retrying
     parsed_s3_url = parse_s3_url(s3_url)
     s3 = s3_client_cache(None, **s3_client_kwargs)
-    return s3.put_object(
-        Body=body,
-        Bucket=parsed_s3_url.bucket,
-        Key=parsed_s3_url.key,
+    retrying = Retrying(
+        wait=wait_random_exponential(multiplier=1, max=15),
+        stop=stop_after_delay(RETRY_STOP_AFTER_DELAY),
+        retry=retry_if_exception_type(RetryableError),
+    )
+    return retrying(
+        _put_object,
+        s3,
+        body,
+        parsed_s3_url.bucket,
+        parsed_s3_url.key,
     )
 
 
+def _put_object(
+    s3_client, body: Any, bucket: str, key: str, **s3_put_object_kwargs
+) -> Dict[str, Any]:
+    try:
+        return s3_client.put_object(
+            Body=body, Bucket=bucket, Key=key, **s3_put_object_kwargs
+        )
+    except ClientError as e:
+        if e.response["Error"]["Code"] == "NoSuchKey":
+            # s3fs may swallow S3 errors - we were probably throttled
+            raise RetryableError(f"Retry upload to: {bucket}/{key}") from e
+        raise NonRetryableError(f"Failed table upload to: {bucket}/{key}") from e
+    except BaseException as e:
+        logger.warning(
+            f"Upload has failed for {bucket}/{key}. Error: {e}",
+            exc_info=True,
+        )
+        raise e
+
+
 def download(
     s3_url: str, fail_if_not_found: bool = True, **s3_client_kwargs
 ) -> Optional[Dict[str, Any]]:
 
-    # TODO (pdames): add tenacity retrying
     parsed_s3_url = parse_s3_url(s3_url)
     s3 = s3_client_cache(None, **s3_client_kwargs)
+    retrying = Retrying(
+        wait=wait_random_exponential(multiplier=1, max=15),
+        stop=stop_after_delay(RETRY_STOP_AFTER_DELAY),
+        retry=retry_if_exception_type(RetryableError),
+    )
+    return retrying(
+        _get_object,
+        s3,
+        parsed_s3_url.bucket,
+        parsed_s3_url.key,
+        fail_if_not_found=fail_if_not_found,
+    )
+
+
+def _get_object(s3_client, bucket: str, key: str, fail_if_not_found: bool = True):
     try:
-        return s3.get_object(
-            Bucket=parsed_s3_url.bucket,
-            Key=parsed_s3_url.key,
+        return s3_client.get_object(
+            Bucket=bucket,
+            Key=key,
         )
     except ClientError as e:
+        if e.response["Error"]["Code"] in TIMEOUT_ERROR_CODES:
+            # Timeout error not caught by botocore
+            raise RetryableError(f"Retry get object from: {bucket}/{key}") from e
+
         if fail_if_not_found:
-            raise
+            raise NonRetryableError(f"Failed get object from: {bucket}/{key}") from e
         else:
             if e.response["Error"]["Code"] != "404":
                 if e.response["Error"]["Code"] != "NoSuchKey":
```

Comment:
resppnse[Error][Code] can either be 404 or NoSuchKey. I think it's latter as this issue should be resolved now: https://github.com/boto/boto3/issues/2442. Otherwise, link this issue for future readers.

Reply:
From throttling tests, I've only seen NoSuchKey pop up, so opted for that.